### PR TITLE
Better exception message when users.xml cannot be loaded due to bad password hash

### LINKED ISSUE #24126: Unexpected behaviour when bad hash in password_sha256_hex
**Describe the unexpected behaviour**
If, when creating a user via xml in the password_sha256_hex field, enter an incorrect hash, then ClickHouse will crash without correct error description

**How to reproduce**
```
<yandex>
    <users>
        <testuser>
           <password_sha256_hex>a95aff94b1f06ec175c359cd1df425c8bb53a1c9c50596ceed8ecbdba9c5752r</password_sha256_hex>
           <profile>readonly</profile>
        </testuser>
    </users>
</yandex>
```

**Expected behavior**
Some human readable error in logs. Maybe just skip this user without crash.

**Error message and/or stacktrace**
/var/log/clickhouse-server/clickhouse-server.log
```
2021.05.14 17:26:08.911812 [ 25348 ] {} <Information> Application: Will watch for the process with pid 25350
2021.05.14 17:26:08.911957 [ 25350 ] {} <Information> Application: Forked a child process to watch
2021.05.14 17:26:08.912391 [ 25350 ] {} <Information> SentryWriter: Sending crash reports is disabled
2021.05.14 17:26:08.912513 [ 25350 ] {} <Trace> Pipe: Pipe capacity is 1.00 MiB
2021.05.14 17:26:08.966801 [ 25350 ] {} <Information> : Starting ClickHouse 21.4.3.21 with revision 54449, build id: 4E8F6FDA87D4EEA23848BE55E4AFF2B2E87C8553, PID 25350
2021.05.14 17:26:08.966941 [ 25350 ] {} <Information> Application: starting up
2021.05.14 17:26:09.092420 [ 25350 ] {} <Information> Application: Calculated checksum of the binary: 1C0FC2BE13EADA0AAFC7B8FB103A0FE8, integrity check passed.
2021.05.14 17:26:09.092580 [ 25350 ] {} <Trace> Application: Will do mlock to prevent executable memory from being paged out. It may take a few seconds.
2021.05.14 17:26:09.105372 [ 25350 ] {} <Trace> Application: The memory map of clickhouse executable has been mlock'ed, total 186.13 MiB
2021.05.14 17:26:09.105515 [ 25350 ] {} <Debug> Application: rlimit on number of file descriptors is 500000
2021.05.14 17:26:09.105529 [ 25350 ] {} <Debug> Application: Initializing DateLUT.
2021.05.14 17:26:09.105537 [ 25350 ] {} <Trace> Application: Initialized DateLUT with time zone 'Europe/Moscow'.
2021.05.14 17:26:09.105556 [ 25350 ] {} <Debug> Application: Setting up /var/lib/clickhouse/tmp/ to store temporary data in it
2021.05.14 17:26:09.105657 [ 25350 ] {} <Debug> Application: Skipped file in temporary path /var/lib/clickhouse/tmp/data
2021.05.14 17:26:09.106004 [ 25350 ] {} <Debug> Application: Configuration parameter 'interserver_http_host' doesn't exist or exists and empty. Will use '***' as replica host.
2021.05.14 17:26:09.106179 [ 25350 ] {} <Information> SensitiveDataMaskerConfigRead: 1 query masking rules loaded.
2021.05.14 17:26:09.106975 [ 25350 ] {} <Debug> ConfigReloader: Loading config '/etc/clickhouse-server/config.xml'
2021.05.14 17:26:09.110256 [ 25350 ] {} <Debug> ConfigReloader: Loaded config '/etc/clickhouse-server/config.xml', performing update on configuration
2021.05.14 17:26:09.111302 [ 25350 ] {} <Information> Application: Setting max_server_memory_usage was set to 44.23 GiB (49.14 GiB available * 0.90 max_server_memory_usage_to_ram_ratio)
2021.05.14 17:26:09.112183 [ 25350 ] {} <Debug> ConfigReloader: Loaded config '/etc/clickhouse-server/config.xml', performed update on configuration
2021.05.14 17:26:09.115887 [ 25350 ] {} <Debug> ConfigReloader: Loading config '/etc/clickhouse-server/users.xml'
2021.05.14 17:26:09.116431 [ 25350 ] {} <Debug> ConfigReloader: Loaded config '/etc/clickhouse-server/users.xml', performing update on configuration
2021.05.14 17:26:09.118825 [ 25350 ] {} <Error> Application: std::exception
2021.05.14 17:26:09.118859 [ 25350 ] {} <Information> Application: shutting down
2021.05.14 17:26:09.118870 [ 25350 ] {} <Debug> Application: Uninitializing subsystem: Logging Subsystem
2021.05.14 17:26:09.119059 [ 25351 ] {} <Trace> BaseDaemon: Received signal -2
2021.05.14 17:26:09.119101 [ 25351 ] {} <Information> BaseDaemon: Stop SignalListener thread
2021.05.14 17:26:09.138296 [ 25348 ] {} <Information> Application: Child process exited normally with code 70.
```

/var/log/clickhouse-server/clickhouse-server.err.log
```
2021.05.14 17:26:09.118825 [ 25350 ] {} <Error> Application: std::exception
```

**Additional context**
ClickHouse version: 21.4.3.21
OS: Ubuntu 16.04

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Better exception message when `users.xml` cannot be loaded due to bad password hash. This closes #24126.

### CODE REVIEW

File: src/Access/Authentication.cpp [hunk 1]
```diff
@@ -9,13 +9,215 @@
...
+    Digest digest;
+    digest.resize(hash.size() / 2);
+
+    try
+    {
+        boost::algorithm::unhex(hash.begin(), hash.end(), digest.data());
+    }
+    catch (const std::exception &)
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Cannot read password hash in hex, check for valid characters [0-9a-fA-F] and length");
```
Comment (alexey-milovidov) [inline src/Access/Authentication.cpp:139]: This.

**Comment by alexey-milovidov** [comment]
It looks like this:
`DB::Exception: Cannot read password hash in hex, check for valid characters [0-9a-fA-F] and length: while parsing user 'default' in users configuration file: while loading configuration file 'users.xml'`

**Comment by alexey-milovidov** [comment]
While doing this task I found tech debt that was left in https://github.com/ClickHouse/ClickHouse/commit/291b7c277acddef298f9df7b8e4e1cca286ce18b

But we don't allow tech debt.

**Review by vitlibar — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@vitlibar I need your help...

**Comment by vitlibar** [comment]
> @vitlibar I need your help...

Yes, sure, I'll help.

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Comment by vitlibar** [comment]
I'll finish the work in https://github.com/ClickHouse/ClickHouse/pull/31557

**Pull request closed without merging**